contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    env, fs,
+    io::{self, Read},
+    path::Path,
+    time::{Duration, Instant},
+};
+
+const MAGIC: [u8; 4] = [0xf9, 0xbe, 0xb4, 0xd9];
+
+fn main() -> Result<(), io::Error> {
+    let start = Instant::now();
+
+    let blocks_path = if env::args().len() > 1 {
+        env::args().nth(1).expect("the arg to exist").into()
+    } else {
+        #[allow(deprecated)]
+        env::home_dir()
+            .expect("to have a home directory")
+            .join(match env::consts::OS {
+                "macos" => "Library/Application Support/Bitcoin",
+                "windows" => "AppData\\Local\\Bitcoin",
+                "linux" => ".bitcoin",
+                _ => {
+                    println!("Unknown OS");
+                    return Ok(());
+                }
+            })
+            .join("blocks")
+    };
+
+    let paths = fs::read_dir(&blocks_path)?
+        .map(|res| res.map(|e| e.path()))
+        .collect::<Result<Vec<_>, io::Error>>()?;
+
+    let xor_path: std::path::PathBuf = blocks_path.join("xor.dat");
+    if !fs::exists(&xor_path)? {
+        println!("No xor.dat file. Make sure you are running Bitcoin Core v28 or higher.");
+        return Ok(());
+    }
+
+    println!("Xor'ing blocks dir. Do not start bitcoind until finished!");
+
+    let key: [u8; 8] = fs::read(&xor_path)?
+        .try_into()
+        .expect("xor.dat to be 8 bytes");
+    if key[0..4] == [0u8; 4] && key[4..] != [0u8; 4] {
+        println!("This script doesn't work with a non-zero key with 4 bytes of leading zeros");
+        return Ok(());
+    }
+    let key = if key == [0u8; 8] {
+        loop {
+            let key: [u8; 8] = rand::random();
+            // Don't use keys with 4 bytes of leading zeros
+            // They won't let us detect the first 4 bytes of magic in the files
+            if key[0..4] == [0u8; 4] {
+                continue;
+            }
+            break key;
+        }
+    } else {
+        key
+    };
+
+    fs::write(xor_path, key)?;
```

Comment:
BTW, this is not atomic. Does it matter? I don't really know. Ideally the tool could continue if it fails but that's not easy.
Still, at least making sure this one doesn't get corrupted may be important?

Reply:
Fixed, we don't overwrite the key now if we didn't create a new one. We also call `sync_data` after writing the new random one.